### DISCUSSION TITLE
Dependency Management in Service Layer: Choosing Between Direct Repository Access and Inter-Service Dependency

### DISCUSSION THREAD

**Discussion by soobinJung** [soobinJung/JCK-World#22] [Architecture and Design] [answered]
안녕하세요,

현재 저는 서비스 계층에서 시험 정보와 관련된 문제들을 조회하기 위해 두 개의 레파지토리에 접근하는 로직을 구현하고 있습니다. 구현에 있어서 두 가지 방법 사이에서 선택해야 하는 상황입니다.

### 1. Repository 간의 Dependency

ExaminationQuestionService 서비스에서 ExaminationQuestionRepository와 ExaminationRepository에 직접 의존하도록 구현합니다. 이 접근 방식은 서비스가 데이터를 직접 조회하며, 각 레파지토리로부터 필요한 데이터를 직접적으로 가져옵니다.

#### 🔊장점

직접적인 의존성
- 서비스가 필요한 데이터를 직접 조회하기 때문에, 데이터 접근 로직이 명확합니다. 이 방법은 중간에 다른 서비스의 비즈니스 로직을 거치지 않으므로, 데이터 접근에 대한 제어가 용이합니다.

순환 참조 위험 감소
- 다른 서비스에 의존하지 않기 때문에, 서비스 간 순환 참조의 가능성이 낮아집니다.

#### 🔊단점

코드 중복
- 이미 ExaminationService에서 Examination 객체를 조회하는 로직이 구현되어 있다면, 이 로직을 ExaminationQuestionService에서도 반복하게 됩니다. 이는 코드 중복을 일으킬 수 있으며, 유지보수성을 저하시킬 수 있습니다.

```
@RequiredArgsConstructor
@Service
public class ExaminationQuestionService {

    private final ExaminationQuestionRepository examinationQuestionRepository;
    private final ExaminationRepository examinationRepository;

    public ResExaminationQuestion findAllByExaminationId(Long examinationId) {
        examinationRepository.findById(id).orElseThrow( () -> new CommonException(CommonExceptionEnum.NO_DATA) );
        // ... 메서드 구현
    }
}
```

### 2. Service 간의 Dependency
ExaminationQuestionService가 ExaminationQuestionRepository와 별도의 ExaminationService에 의존하도록 구현할 수 있습니다.
이 경우, ExaminationService는 중복 로직을 피하기 위해 ExaminationRepository의 로직을 재사용합니다.

#### 🔊 장점

코드 재사용
- ExaminationService 내의 Examination 조회 로직을 재사용함으로써 코드 중복을 줄일 수 있습니다. 이는 유지보수성을 향상시키고, 코드의 일관성을 유지하는 데 도움이 됩니다.

추상화된 접근
- Examination 데이터에 접근할 때 ExaminationService를 통하면, 해당 서비스 내의 추가적인 비즈니스 로직이나 검증 로직을 활용할 수 있습니다. 이는 데이터 처리의 일관성과 안정성을 높일 수 있습니다.

#### 🔊단점

순환 참조 위험
- 서비스 간의 의존성이 복잡해지면, 특히나 크고 복잡한 애플리케이션에서는 순환 참조의 위험이 증가합니다. 순환 참조는 애플리케이션의 시작 실패를 초래할 수 있으며, 디버깅이 어려울 수 있습니다.

```
@RequiredArgsConstructor
@Service
public class ExaminationQuestionService {

    private final ExaminationQuestionRepository examinationQuestionRepository;
    private final ExaminationService examinationService;

    public ResExaminationQuestion findAllByExaminationId(Long examinationId) {
       examinationService.getExaminationById(examinationId)
        // ... 메서드 구현
    }
}
```

```
@RequiredArgsConstructor
@Service
public class ExaminationService {

    private final ExaminationRepository examinationRepository;

    public List<Examination> getExamination() {
        return examinationRepository.findAll();
    }

    public Examination getExaminationById(Long id) {
        return examinationRepository.findById(id).orElseThrow( () -> new CommonException(CommonExceptionEnum.NO_DATA) );
    }
}
```

**Comment by jincrates** [comment, marked as answer]
레이어드 아키텍처로 개발을 하신거 같은데, 각 레이어 간의 의존규칙을 정하고 개발을 하면 좋을 것 같아요! 
레이어간 규칙에 대해서는 다음 영상 8:44 정도부터 보시면 좋을 것 같습니다.
https://youtu.be/RVO02Z1dLF8?t=524

**Reply by soobinJung** [reply]
layered architecture 에 대해 연구해봐야겠네요 .. !!
감사합니다 !!
